framework/include/libtorch/controls/LibtorchNeuralNetControl.h

```diff
@@ -0,0 +1,100 @@
+  LibtorchNeuralNetControl(const InputParameters & parameters);
+
+  /// Execute neural network to determine the controllable parameter values
+  virtual void execute() override;
+
+  /// Get the (signal_index)-th signal of the control neural net
+  Real getSignal(const unsigned int signal_index);
+
+  /// Get the number of controls this object is computing
+  unsigned int numberOfControlSignals() { return _control_names.size(); }
+
+  /**
+   * Function responsible for loading the neural network for the controller. This function is used
+   * when copying the neural network from a main app which trains it.
+   */
+  void loadControlNeuralNet(const std::shared_ptr<Moose::LibtorchArtificialNeuralNet> & input_nn);
+
+  /// Return a pointer to the stored neural networt
+  const Moose::LibtorchNeuralNetBase & controlNeuralNet() const;
+
+protected:
+  /// Function responsible for checking for potential user errors in the input file
+  void conditionalParameterError(const std::string & param_name,
+                                 const std::vector<std::string> & conditional_param,
+                                 bool should_be_defined = true);
+
+  /// Function that updates the values of the current response
+  void updateCurrentResponse();
+
+  /// Function that prepares the input tensor for the controller neural network
+  torch::Tensor prepareInputTensor();
+
+  /// The values of the current observed postprocessor values
+  std::vector<Real> _current_response;
+  /// This variable is populated if the controller needs acess to older values of the
+  /// observed postprocessor values
+  std::vector<std::vector<Real>> _old_responses;
+
+  /// The names of the controllable parameters
+  const std::vector<std::string> & _control_names;
+  /// The control signals from the last evaluation of the controller
+  std::vector<Real> _current_control_signals;
+
+  /// Names of the postprocessors which contain the observations of the system
+  const std::vector<PostprocessorName> & _response_names;
+
+  /// Links to the current response postprocessor values. This is necessary so that we can check
+  /// if the postprocessors exist.
+  std::vector<const Real *> _response_values;
+
+  /// Number of timesteps to use as input data from the reporters (this influences how many past
+  /// results are used, e.g. the size of _old_responses)
+  const unsigned int _input_timesteps;
+
+  /// Flag to show if the vector containing the time-series of responses has been initialized or not
+  bool _initialized;
+
+  /// Shifting constants for the responses
+  const std::vector<Real> _response_shift_factors;
+  /// Scaling constants (multipliers) for the responses
+  const std::vector<Real> _response_scaling_factors;
+  /// Multipliers for the actions
+  const std::vector<Real> _action_scaling_factors;
+
+  /// Pointer to the neural net object which is supposed to be used to control
+  /// the parameter values
+  std::shared_ptr<Moose::LibtorchNeuralNetBase> _nn;
```

Comment:
```suggestion
  Moose::LibtorchNeuralNetBase * _nn;
```

Reply:
This object will own the NN, and the `shared_ptr` is important for the file-reading capabilities. This guy is either populated by a fileIO (`torch::load`) or by a transfer which uses a copy constructor.

Reply:
I need the shared pointer for `torch::load` and `torch::save`. With the controller returning a reference this guy is the sole owner.

Reply:
The transfer will copy construct a neural net in this object.